### DISCUSSION TITLE
How to determine whether `getServerSideProps` was triggered?

### DISCUSSION THREAD

**Discussion by mkotsollaris** [vercel/next.js#47284] [Help]
### Summary

In my application, I am using `getServerSideProps`, which works great, but for browser navigation my state management is quite hacky as I need to "invalidate" my client data and be able to re-compute certain elements to re-render the page.

Is there a more native-approach to manage this? I find that if I had a lifecycle hook/function on the client that I know triggers everytime `getServerSideProps` occurs, that would help with this.

When using `getInitialProps` this wasn't too difficult to achieve, as we would call `getInitialProps` on the client and force-invalidate whatever state we need. However, we can't do the same with `getServerSideProps` being a server-only function.

I thought that something along these lines would help:

```tsx
export async function getServerSideProps(ctx: GetServerSidePropsContext) {
   return {
      props: { value: Math.random()} 
  }
}

export const MyPage: NextPage = (initialProps) => {
  const [clientProps, setClientProps] = useState(initialProps);
  
  // this is the `getServerSideProps` lifecycle Hook which I want to re-compute my data on the client
  const onGetServerSidePropTrigger = (props) => {
    setClientProps(props)
  }

  return <> ....</>
}
```

Right now, identifying when props change from `getServerSideProps` is not as straightforward and requires a combination of `useEffect` using `useRouter` to achieve this, I feel there should be a more straightforward way to do this via a native hook or something along these lines.

I suppose what I am after is a reliable way to determine that `getServerSideProps` was called on the client so that I perform a certain computation down in the state tree 🤔, but I am not sure what's the best way to go around this.

**Comment by icyJoseph** [comment]
Hi,

GSSP runs exclusively server side. 

There's two situations, roughly speaking:

- the user arrives to your application, at a GSSP page
   - Then GSSP runs server side and its result is used to generate an entire HTML page, which is then sent to the client
- the user is already in your app, and navigates, using next/link or next/router to a GSSP page:
   - then, as soon as you transition, Next.js makes a request to the server, `_next/data`, where GSSP runs server side, and the results are returned as JSON, which is then parsed as a JS object and used as props for your page

So I am not sure about what you are after here. You can trigger a GSSP call, without leaving the page: https://www.joshwcomeau.com/nextjs/refreshing-server-side-props/

But if the page itself uses GSSP, and you navigate away from it and come back, you'd literally get new props by design. 

Even when you use a dynamic route, to load many pages under the same page, you don't need to manage a "GSSP-life-cycle" - https://github.com/vercel/next.js/discussions/45077#discussioncomment-4737341

Could you explain a bit more of your situation?

**Reply by mkotsollaris** [reply]
In my application, which is a typical search-result page (you can think of something similar to [google flights](https://www.google.com/travel/flights)), quite a few state updates happen during user interactions that run purely on the client. As such, `Router.push(shallow)` is used to avoid triggering `GSSP` in these scenarios. During certain events, like browser navigation (forward/back btn) we want to trigger GSSP. My logic is basically detecting these events and 'overriding' the shallow property to `false` in order to 'force' a trigger on `GSSP` in these cases through this `useEffect`:

```jsx
  useEffect(() => {
    router.beforePopState(({ url, as, options }) => {
      if (options.shallow) {
        router.replace(url, as, { shallow: false });
      }
      return true;
    });

    return () => {
      router.beforePopState(() => true);
    };
  }, []);
```

However, there are a few other cases that I trigger `GSSP` in my app, and I can't find a reliable way of telling whether there was a trigger of `GSSP` in my app (client). 

Hopefully that helps showcase why I thought a hook/event like having a hook `onGetServerSidePropTrigger` would be handy.


<img width="782" alt="image" src="https://user-images.githubusercontent.com/8595514/227308607-f8a7920a-ed4b-4b60-b34d-30d6d6c484d5.png">

**Reply by chiralium** [reply]
Hello. I have catalog page with some products and filtering. Filter parameters adding to url by `Router.push(shallow)` (so we not trigger GSSP), but after that, if I hit "Back Button" sometimes `GSSP` trigger, sometimes not.  Did you know why?
